internal/configs/transportserver.go

```diff
@@ -59,14 +59,17 @@ func generateTransportServerConfig(transportServerEx *TransportServerEx, listene
 		proxyTimeout = transportServerEx.TransportServer.Spec.SessionParameters.Timeout
 	}
 
-	statusZone := ""
+	serverSnippets := generateSnippets(snippetsEnabled, transportServerEx.TransportServer.Spec.ServerSnippets, []string{})
+	if !snippetsEnabled && (transportServerEx.TransportServer.Spec.ServerSnippets != "") {
+		return nil, fmt.Errorf("snippet specified but snippets feature is not enabled")
```

Comment:
when we have a warning, we don't stop processing a resource - we record a warning and continue. Could you not return an error here, as it is not an error? This would be similar to how we handle warnings for VirtualServer 

I suggest removing this check from this PR completely and address it (with unit tests) in the next PR when warnings will be propagated fully to the Controller 

sorry, I just noticed this change :(

Reply:
I think this is an error. If you specify a snippets in a resource, then you expect them to be enabled and working, otherwise why would the user add them?

Reply:
Yes i agree with Sean. If the user tries to use a snippet and the feature isn't enabled then that means the resource doesn't work and is invalid. We don't want to deploy half a resource or a resource with half the functionality that the user defined in the yaml configuration.

Reply:
it makes sense: better to report an error and reject a TS in this case. This would be different from VS, but it's better user experience.

however, such validation happens here https://github.com/nginxinc/kubernetes-ingress/blob/master/pkg/apis/configuration/validation/transportserver.go#L25 not in `generateTransportServerConfig`.

Reply:
yes but snippets don't get validated and we don't provide such data, like if a command line arg is set, to that validation process.

Reply:
yeah, but if you want users to get an error when they use snippets when snippets are disabled, validation package is the right place - this is where we do validation. that will involve propagating enable snippets status to transport server validator

Reply:
Users do get an error when they use snippets when snippets are disabled. The `generateTransportServerConfig` validation error is propagated to `updateTransportServerEvents` function in `controller.go` which will throw an event that will notify the user. 

This validation of `snippets` requires the k8s resource and the context of the IC. This package handles the validation of the k8s resource fields and it is not like the resource is not formed correctly or has incorrect fields. It's a valid string and thats all the context this package should have. The CLI arg is an implementation detail and is independent of the resource so this validation package shouldn't know about it.

Reply:
The added approach with handling errors is inconsistent with the approach we use for config generation: we always rely on the latest version of the resource, not the latest good version.

When a resource is rejected with an error, but its config is not removed or regenerated, this means an IC replica keeps using the latest good version.  However, other replicas might have a different good version stored. More over, new replicas (which could be created in case of replicas failures, node draining, scaling, upgrades) will only see the latest version in the cluster. As a result, the configuration will skew among replicas.

> This package handles the validation of the k8s resource fields and it is not like the resource is not formed correctly or has incorrect fields.

the validation depends on the enabled features. For example, validation for Policies and VirtualServers rejects resources that use Plus features, validation for TransportServers rejects TLS Passthrough resources if TLS Passthrough is not enabled.

Note that Plus or Passthrough are configured via cli args - so they are constants, because changing CLI arguments requires an upgrade of a deployment - it is not something that could be configured via ConfigMap or GlobalConfiguration. That's why it is possible to make validation decisions based on those conditions.

if you don't think a configured snippet is an error, then it should be a warning and handled like a warning - in that case the config is still generated, but the warning is reported to the users as an event and in the status of the resource.

Reply:
Think I have it:
```
Status:
  Message:  TransportServer default/dns-tcp was rejected with error: spec.serverSnippets: Forbidden: snippet specified but snippets feature is not enabled
  Reason:   Rejected
  State:    Invalid
Events:
  Type     Reason    Age                From                      Message
  ----     ------    ----               ----                      -------
  Warning  Rejected  20s (x2 over 21s)  nginx-ingress-controller  TransportServer default/dns-tcp was rejected with error: spec.serverSnippets: Forbidden: snippet specified but snippets feature is not enabled
```